rust/agama-lib/src/scripts/model.rs

```diff
@@ -0,0 +1,215 @@
+#[strum(serialize_all = "camelCase")]
+#[serde(rename_all = "camelCase")]
+pub enum ScriptsGroup {
+    Pre,
+    Post,
+}
+
+/// Represents a script to run as part of the installation process.
+#[derive(Clone, Serialize, Deserialize)]
+pub struct Script {
+    /// Script's name.
+    pub name: String,
+    /// Script's body. Either the body or the URL must be specified.
+    pub body: Option<String>,
+    /// URL to get the script from. Either the body or the URL must be specified.
+    pub url: Option<String>,
+    /// Script's group
+    pub group: ScriptsGroup,
+}
+
+impl Script {
+    /// Runs the script and returns the output.
+    ///
+    /// * `workdir`: where to write assets (script, logs and exit code).
+    pub async fn run(&self, workdir: &Path) -> Result<(), ScriptError> {
+        let dir = workdir.join(self.group.to_string());
+
+        let path = dir.join(&self.name);
+        self.write(&path).await?;
+
+        let output = process::Command::new(&path).output()?;
+
+        let stdout_log = dir.join(format!("{}.log", &self.name));
+        fs::write(stdout_log, output.stdout)?;
+
+        let stderr_log = dir.join(format!("{}.err", &self.name));
+        fs::write(stderr_log, output.stderr)?;
+
+        let status_file = dir.join(format!("{}.out", &self.name));
+        fs::write(status_file, output.status.to_string())?;
+
+        Ok(())
+    }
+
+    /// Writes the script to the file system.
+    ///
+    /// * `path`: path to write the script to.
+    async fn write<P: AsRef<Path>>(&self, path: P) -> Result<(), ScriptError> {
+        let mut file = fs::OpenOptions::new()
+            .create(true)
+            .write(true)
+            .truncate(true)
+            .mode(0o500)
+            .open(&path)?;
+
+        if let Some(url) = &self.url {
+            Transfer::get(url, file)?;
+        } else if let Some(body) = &self.body {
+            write!(file, "{}", &body)?;
+        }
+        // FIXME: else: invalid script definition
+
+        Ok(())
+    }
+}
+
+/// Manages a set of installation scripts.
+///
+/// It offers an API to add and execute installation scripts.
+pub struct ScriptsRepository {
+    workdir: PathBuf,
+    pub scripts: Vec<Script>,
+}
+
+impl ScriptsRepository {
+    /// Builds a new repository.
+    ///
+    /// * `workdir`: directory to store the scripts.
+    pub fn new<P: AsRef<Path>>(workdir: P) -> ScriptsRepository {
+        ScriptsRepository {
+            workdir: PathBuf::from(workdir.as_ref()),
+            ..Default::default()
+        }
+    }
+
+    /// Adds a new script to the repository.
+    ///
+    /// * `script`: script to add.
+    pub fn add(&mut self, script: Script) {
+        self.scripts.push(script);
+    }
+
+    /// Removes all the scripts from the repository.
+    pub fn clear(&mut self) {
+        self.scripts.clear();
+    }
+
+    /// Runs the scripts in the given group.
+    ///
+    /// They run in the order they were added to the repository.
```

Comment:
I would document here that it does not return error if script failed. and in this case just log error.